### PR TITLE
Added CSRF validation (closes #362, #375)

### LINKED ISSUE #362: [DX] Add csrf_provider in @Security annotation ?
Could be like :
```
/*
*@Security("isCsrfTokenValid('secret',csrf)")
*@Route("/message/{id}/remove/{csrf}", name="message_remove")
*/
public function removeMessageAction(Message $message, $csrf){
[...]
}
```

### PR DESCRIPTION
**Pull request by Koc**
Closes #362, #375 

TODO:
- [x] Add tests

### CODE REVIEW

File: Resources/config/csrf.xml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+    xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="sensio_framework_extra.security.listener" class="Sensio\Bundle\FrameworkExtraBundle\EventListener\CsrfListener">
+            <argument type="service" id="security.csrf.token_manager" on-invalid="null" />
```
Comment (Koc) [inline Resources/config/csrf.xml]: Can I make it required and validate existence when creating new instance?

**Comment by Koc** [comment]
@fabpot tests pass. Please review

File: EventListener/CsrfListener.php [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        /* @var $configuration Csrf */
+
+        if (null === $this->csrfTokenManager) {
+            throw new \LogicException('To use the @Csrf tag, you need to install the "symfony/security-csrf".');
+        }
+
+        $token = $request->get($configuration->getParam());
+
+        if (!$this->csrfTokenManager->isTokenValid(new CsrfToken($configuration->getIntention(), $token))) {
+            throw new BadRequestHttpException('Csrf token is invalid.');
```
Comment (jakzal) [inline EventListener/CsrfListener.php]: I found examples of both. [Spring returns a 403](http://docs.spring.io/spring-security/site/docs/current/reference/html/csrf.html#csrf-configure):

> The next step is to include Spring Security’s CSRF protection within your application. Some frameworks handle invalid CSRF tokens by invaliding the user’s session, but this causes its own problems. Instead by default Spring Security’s CSRF protection will produce an HTTP 403 access denied.

File: Configuration/Csrf.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+ * @author Konstantin Myakshin <koc-dp@yandex.ru>
+ *
+ * @Annotation
+ * @Target("METHOD")
+ */
+class Csrf extends ConfigurationAnnotation
+{
+    protected $intention;
+
+    protected $param = '_token';
```
Comment (fabpot) [inline Configuration/Csrf.php]: Can we make the properties private?

File: EventListener/CsrfListener.php (hunk 2 above)
Comment (fabpot) [inline EventListener/CsrfListener.php]: I think I would slightly prefer the AccessDenied one.

File: Resources/doc/annotations/csrf.rst [hunk 4]
```diff
@@ -0,0 +1,38 @@
+@Security
```
Comment (fabpot) [inline Resources/doc/annotations/csrf.rst]: `@Csrf`?

File: Resources/doc/annotations/csrf.rst [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+=========
+
+.. caution::
+
+    The ``@Csrf`` annotation only works as of Symfony 2.7.
```
Comment (fabpot) [inline Resources/doc/annotations/csrf.rst]: `The ``@Csrf`` annotation requires Symfony 2.7+.`?

File: Resources/doc/annotations/csrf.rst [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+         */
+        public function deleteAction()
+        {
+            // ...
+        }
+    }
+
+On the template side
+
+    <a href="{{ path('acme_post_delete', {'_token': csrf_token('post_delete')}) }}">delete post</a>
```
Comment (fabpot) [inline Resources/doc/annotations/csrf.rst]: I would prefer that you use an example that does not use GET as it is really a bad practice that we should not encourage.

File: Tests/EventListener/CsrfListenerTest.php [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+{
+    protected $listener;
+
+    protected $tokenManager;
+
+    protected function setUp()
+    {
+        if (!interface_exists('Symfony\Component\Security\Csrf\CsrfTokenManagerInterface')) {
+            $this->markTestSkipped('This test requires "symfony/security-csrf".');
+        }
```
Comment (fabpot) [inline Tests/EventListener/CsrfListenerTest.php]: This should be removed as we list all required deps in composer.json so it should always be present.

File: Tests/EventListener/CsrfListenerTest.php [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+        $this->tokenManager = $this
+            ->getMockBuilder('Symfony\Component\Security\Csrf\CsrfTokenManagerInterface')
+            ->getMock();
+
+        $this->listener = new CsrfListener($this->tokenManager);
```
Comment (fabpot) [inline Tests/EventListener/CsrfListenerTest.php]: Please remove the `setUp()` method and the property and creates instances directly in tests. You can of course create helper methods like you did for the Request and the Event.

**Comment by Koc** [comment]
update:
* annotation properties maked private
* added default intention equal to `default`
* updated docs
* updated tests
* created own exception. I doubt that reusing AccessDeniedException is good idea because of it depends on attributes and subject and it was created for authorization errors. Also this new exception allow pass valid token - useful for handling invalid tokens on client - we can refresh token, ask user and allow retry last action with new token.

**Comment by Koc** [comment]
tests failures looks like unrelated to this PR

File: Resources/config/csrf.xml (hunk 1 above)
Comment (xabbuh) [inline Resources/config/csrf.xml]: Why not simply removing this definition if the `security.csrf.token_manager` service is not available?

File: EventListener/CsrfListener.php [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+/**
+ * CsrfListener handles csrf restrictions on actions.
+ *
+ * @author Konstantin Myakshin <koc-dp@yandex.ru>
+ */
+class CsrfListener implements EventSubscriberInterface
+{
+    private $csrfTokenManager;
+
+    public function __construct(CsrfTokenManagerInterface $csrfTokenManager = null)
```
Comment (xabbuh) [inline EventListener/CsrfListener.php]: Allowing `null` here makes it possible to create the listener in an unusable state. I think you should rather remove (or not create) the definition instead when no CSRF token manager is available (see also my comment below).

File: Resources/doc/annotations/csrf.rst [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+         */
+        public function batchDeleteAction()
+        {
+            // ...
+        }
+    }
+
+On the template side
+
+    <form action="{{ path('acme_posts_batch_delete') }}" method="POST">
```
Comment (xabbuh) [inline Resources/doc/annotations/csrf.rst]: This should be put into a code block too.

File: EventListener/CsrfListener.php (hunk 9 above)
Comment (Koc) [inline EventListener/CsrfListener.php]: I've just reflect logic from security listener.

File: Resources/doc/annotations/csrf.rst (hunk 10 above)
Comment (HeahDude) [inline Resources/doc/annotations/csrf.rst]: You can use a double colon for PHP code blocks:
```rst
On the template side::

    <form action="{{ path('acme_posts_batch_delete') }}" method="POST">
```

Comment (HeahDude) [inline Resources/doc/annotations/csrf.rst]: Edited my post since the code is html :)

**Comment by Koc** [comment]
rebased on master again

ping @fabpot 

questions that required to be resolved:
1a. should I add and register custom exception handler for converting `InvalidCsrfTokenException` to 403 response? Invalid csrf produces 500 error for now. 
1b. AccessDeniedException is not good idea because reasons described in https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/430#issuecomment-249187438
1c. Or extends from `HttpException` from http-kernel?
2. my listener reflects logic from security listener. But is it possible check availability of the required services inside configuration and enable/disable required annotations depends on services availability. Like https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L512 but we need check services.

**Comment by Koc** [comment]
@fabpot can you make review and give advices to my previous questions, please? Would be nice finish this PR.

**Comment by fabpot** [comment]
Closing as this repository is not maintained anymore. As of Symfony 6.2, all features have been moved to Symfony core now.
See #783

**Pull request closed without merging**